src/gpu_sysman.c

```diff
@@ -768,16 +771,44 @@ static int gpu_init(void) {
 }
 
 /* Add device labels to all metrics in given metric family and submit family to
- * collectd.  Resets metric family after dispatch */
+ * collectd, and log the metric if metric logging is enabled.
+ * Resets metric family after dispatch */
 static void gpu_submit(gpu_device_t *gpu, metric_family_t *fam) {
+  const char *pci_bdf = gpu->pci_bdf;
+  struct timespec ts = {0, 0};
+  if (config.logmetrics) {
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    /* skip common BDF address prefix */
+    if (strncmp("0000:", pci_bdf, 5) == 0) {
+      pci_bdf += 5;
+    }
+  }
   metric_t *m = fam->metric.ptr;
-  for (size_t i = 0; i < fam->metric.num; i++) {
-    metric_label_set(m + i, "pci_bdf", gpu->pci_bdf);
+  for (size_t i = 0; i < fam->metric.num; m++, i++) {
+    /* log metric values in addition to dispatching them? */
+    if (config.logmetrics) {
+      const char *type = "<type>";
+      for (size_t l = 0; l < m->label.num; l++) {
+        const char *name = m->label.ptr[l].name;
+        /* labels list is sorted, so e.g. "direction" will match before "type"
+         */
+        if (strcmp(name, "direction") == 0 || strcmp(name, "location") == 0 ||
+            strcmp(name, "type") == 0) {
+          type = m->label.ptr[l].value;
+          break;
+        }
+      }
```

Comment:
Great suggestion, applied.

PS. that forced me to drop the mocked metric functions from `gpu_sysman_test.c`, and include real one instead.

FYI: I've been a bit hesitant on using collectd's own functions in plugin tests instead of mocked ones, as their dependencies pull in other collected code, which fails to build with `gcc -Werror` + the extra compiler warnings [1] I'm always using for the test code.  That's why I do not use collectd build scripts for testing, and `#include` relevant collectd source files directly in the test code.

[1] For a subset of those, see e.g. https://github.com/collectd/collectd/pull/3970 and https://github.com/collectd/collectd/pull/3917.

Reply:
> that forced me to drop the mocked metric functions from gpu_sysman_test.c, and include real one instead.

While above mentioned `-Werror` tests passed with GCC in Fedora 38, they did not pass with Ubuntu 22.04 LTS => added workaround to `src/utils/metadata/meta_data.c`.

Reply:
I would actually recommend to prefer the real implementation over mocked or faked ones – you'll actually test the production code this way. I discovered #4214 when writing unit tests for the Write Prometheus plugin, for example.

Reply:
Do you want me to split the 3 testing related commits to a separate PR?

Reply:
I split independent testing changes to a separate PR https://github.com/collectd/collectd/pull/4223.